### PR TITLE
[Session] fix: session mismatch due to off-by-one block height

### PR DESCRIPTION
**Pull request by red-0ne**
### Human Summary

Fix off-by-one block height causing `session mismatch` errors during relaying.

### AI Summary

reviewpad:summary

## Issue

- #366 

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_develop_and_test`
- [x] **Run E2E tests locally**: `make test_e2e`
- [ ] **Run E2E tests on DevNet**: Add the `devnet-test-e2e` label to the PR. This is VERY expensive, only do it after all the reviews are complete.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: x/session/keeper/keeper.go [hunk 1]
```diff
@@ -60,12 +60,11 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
-	// ctx.BlockHeader().Height is the height of the current block being processed,
-	// decrementing it by 1 gives us the height of the last block committed.
-	height := ctx.BlockHeader().Height - 1
-	// If height is 0, ctx.BlockHeader().LastBlockId.Hash will be nil,
+	// ctx.BlockHeader().Height is the height of the last committed block.
+	height := ctx.BlockHeader().Height
+	// If height is 1, ctx.BlockHeader().LastBlockId.Hash will be nil,
 	// fall back to an empty byte slice.
-	if height == 0 {
+	if height == 1 {
```
Comment (Olshansk) [inline x/session/keeper/keeper.go:67]: Why is this 1?

If the last commit block is 1, it means we have committed a real block. Genesis is 0.

File: x/session/keeper/keeper.go [hunk 2]
```diff
@@ -60,12 +60,11 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 func (k Keeper) BeginBlocker(ctx sdk.Context) {
 	// ctx.BlockHeader().LastBlockId.Hash is the hash of the last block committed
 	hash := ctx.BlockHeader().LastBlockId.Hash
-	// ctx.BlockHeader().Height is the height of the current block being processed,
-	// decrementing it by 1 gives us the height of the last block committed.
-	height := ctx.BlockHeader().Height - 1
-	// If height is 0, ctx.BlockHeader().LastBlockId.Hash will be nil,
+	// ctx.BlockHeader().Height is the height of the last committed block.
+	height := ctx.BlockHeader().Height
```
Comment (Olshansk) [inline x/session/keeper/keeper.go:64]: @red-0ne Can you explain what we missed or misunderstood in our last review cycle? I want to understand why it got through code review.

Screenshot 2024-02-06 at 9 26 08 PM

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: x/session/keeper/keeper.go (hunk 1 above)
Comment (red-0ne) [reply x/session/keeper/keeper.go:67]: Yes, block `1` should behave just like the others, but for some reason it doesn't have its `ctx.BlockHeader().LastBlockId` populated and the `Hash` has its 0 value that panics when the CosmosSDK store checks for valid values.

The empty byte slice seems to fix it (but we can assign it any valid byte slice value)

image

Will have a better comment if you are OK with the solution.

File: e2e/tests/session.feature [hunk 3]
```diff
@@ -5,7 +5,7 @@ Feature: Session Namespace
   Scenario: Supplier completes claim/proof lifecycle for a valid session
     Given the user has the pocketd binary installed
     When the supplier "supplier1" has serviced a session with "5" relays for service "svc1" for application "app1"
-    And the user should wait for "5" seconds
+    And the user should wait for "7" seconds
```
Comment (Olshansk) [inline e2e/tests/session.feature:8]:
```suggestion
    # TODO_TECHDEBT: Once the session grace period is configurable, set it to 0
    # at the beginning of the test to reduce this wait time as much as possible.
    And the user should wait for "7" seconds
```

File: x/session/keeper/keeper.go (hunk 2 above)
Comment (Olshansk) [reply x/session/keeper/keeper.go:64]: Can you #PUC to a link that proves `ctx.BlockHeader().Height is the height of the last committed block.`

File: x/session/keeper/keeper.go (hunk 1 above)
Comment (Olshansk) [reply x/session/keeper/keeper.go:68]: I appreciate the workaround and the detailed comment, but I still don't understand "why". maybe it's a bug in cosmos or rollkit? I want to be convinced that we're not bandainding this but actually understand why its the solution.

For example, see this comment where I reached out to cosmos when I didn't understand something in the past: https://github.com/pokt-network/poktroll/pull/323#discussion_r1463541516

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: x/session/keeper/keeper.go (hunk 2 above)
Comment (red-0ne) [reply x/session/keeper/keeper.go:64]: This has been confirmed by comparing the block heights and hashes we got from the CometBFT websocket client and the values of `ctx.BlockHeader().Height` and `ctx.BlockHeader().LastBlockId.Hash`.

I couldn't find any statement of that in the CosmosSDK or code. The closest I got was this reference:

The only relevant info I found was this https://github.com/cosmos/cosmos-sdk/blob/main/baseapp/grpcserver.go#L60

File: e2e/tests/session.feature (hunk 3 above)
Comment (red-0ne) [reply e2e/tests/session.feature:8]: I don't understand how this relates to session grace period configuration. To me it's the time we need to wait before querying for claims and see if they incremented. I believe it mostly serves as a timeout.

File: x/session/keeper/keeper.go (hunk 1 above)
Comment (red-0ne) [reply x/session/keeper/keeper.go:68]: image

This should explain why there's no hash for block `1`. Will update the comments accordingly.

We are also assuming that genesis starts at block `0`. We should revisit our session start/end calculations and their corresponding tests. Will create a ticket for that.

File: x/session/keeper/query_get_session.go [hunk 4]
```diff
@@ -21,16 +21,14 @@ func (k Keeper) GetSession(goCtx context.Context, req *types.QueryGetSessionRequ
 
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	// If block height is not specified, use the current (context's latest) block height
+	// GetSession should be deterministic and always return the same session for
```
Comment (Olshansk) [inline x/session/keeper/query_get_session.go:24]: Can you please move this into a godoc comment above

File: x/session/keeper/query_get_session.go [hunk 5]
```diff
@@ -21,16 +21,14 @@ func (k Keeper) GetSession(goCtx context.Context, req *types.QueryGetSessionRequ
...
+	// the same block height.
 	// Note that `GetSession` is called via the `Query` service rather than the `Msg` server.
 	// The former is stateful but does not lead to state transitions, while the latter one
 	// does. The request height depends on how much the node has synched and only acts as a read,
 	// while the `Msg` server handles the code flow of the validator/sequencer when a new block
 	// is being proposed.
 	blockHeight := req.BlockHeight
```
Comment (Olshansk) [inline x/session/keeper/query_get_session.go:31]: Consider adding a logigng statement "getting session for height blockHeight"

File: x/session/keeper/keeper.go [hunk 6]
```diff
@@ -60,12 +60,12 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
 	hash := ctx.BlockHeader().LastBlockId.Hash
-	// ctx.BlockHeader().Height is the height of the current block being processed,
-	// decrementing it by 1 gives us the height of the last block committed.
-	height := ctx.BlockHeader().Height - 1
-	// If height is 0, ctx.BlockHeader().LastBlockId.Hash will be nil,
-	// fall back to an empty byte slice.
-	if height == 0 {
+	// ctx.BlockHeader().Height is the height of the last committed block.
+	height := ctx.BlockHeader().Height
+	// When height is 1, ctx.BlockHeader().LastBlockId.Hash will not be populated,
```
Comment (Olshansk) [inline x/session/keeper/keeper.go:65]: @red-0ne Please create a ticket for `We are also assuming that genesis starts at block 0. We should revisit our session start/end calculations and their corresponding tests. Will create a ticket for that.` and add a TODO_TECHDEBT in the code in the appropriate location.

File: x/session/keeper/keeper.go [hunk 7]
```diff
@@ -60,12 +60,12 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
+	// and leads to a panic when trying to save the hash in the store.
```
Comment (Olshansk) [inline x/session/keeper/keeper.go:66]: @red-0ne This is a good comment but if we copy-paste (and reword) what marko told you as a comment, there will be no ambiguity fo anyone.

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Review by Olshansk — APPROVED** [review]

**Pull request merged**